accepted/2020/or-greater-defines/or-greater-defines.md

```diff
@@ -0,0 +1,221 @@
+```
+
+### When running on a later version of .NET, use a newer API
+
+Zachariah is building a .NET imaging library. In .NET 5, we have added enough
+hardware intrinsics so that he can implement the blur functionality in a much
+faster way. However, he still wants to support .NET Standard 2.0, so he decides
+to use `#if` to multi-target for .NET 5.
+
+Since he checks for the minimum version of the framework that introduced the
+API, adding a new version wouldn't change this code, so he simply puts the
+fallback logic for the slow version in the `#else` branch, with no call to
+`#error`.
+
+```C#
+public class Picture
+{
+    public void Blur(float radius)
+    {
+#if NET5_0_OR_GREATER
+        FastBlurUsingHardwareIntrinsics();
+#else
+        SlowBlurUsingSimpleMath();
+#endif
+    }
+}
+```
+
+## Requirements
+
+### Goals
+
+* The new symbols don't change semantics of existing code
+* The new symbols allow expressing a `>=` relationship with version numbers
+* The new symbols apply to both frameworks as well as platforms
+* The new symbols are defined for all frameworks that are relevant for package
+  authors today, which includes:
+    - .NET 5
+    - .NET Core
+    - .NET Framework
+    - .NET Standard
+
+### Non-Goals
+
+* Making existing code more resilient to TFM changes
+
+## Design
+
+The SDK will define `XXX_OR_GREATER` variants for the following TFMs:
+
+* **.NET Framework**
+    - Applies to `net1.0`-`net4x` and the `NETFRAMEWORK`/`NET` symbols
+    - For example, `net4.8` will define `NETFRAMEWORK`, `NET48`,
+      `NET48_OR_GREATER`, `NET472_OR_GREATER`..`NET20_OR_GREATER`.
+* .NET Core
+    - Applies to `netcoreappX.Y` and the `NETCOREAPP` symbol
+    - For example, `netcoreapp3.1` will define `NETCOREAPP`, `NETCOREAPP3_1`,
+      `NETCOREAPP3_1_OR_GREATER`..`NETCOREAPP1_0_OR_GREATER`.
+* .NET 5 and later
+    - Applies to `netX.Y` and the `NET` symbol
+    - For example, `net6.0` will define `NET`, `NET6_0`, `NET6_0_OR_GREATER` and
+      `NET5_0_OR_GREATER`.
+    - This will also include the corresponding defines a .NET Core 3.1 successor
+      would have gotten. For example, `net5.0` will also define `NETCOREAPP`,
+      `NETCOREAPP3_1_OR_GREATER`..`NETCOREAPP1_0_OR_GREATER`.
+    - This will *neither* define a `NETCOREAPP5_0` nor
+      `NETCOREAPP5_0_OR_GREATER`.
+* .NET 5 and later with operating systems
+    - The OS flavors will also gain the `XXX_OR_GREATER` variants
+    - For example, `net5.0-windows10.0.19222.0` will also define `WINDOWS`,
+      `WINDOWS10_0_19041_0`, `WINDOWS10_0_19041_0_OR_GREATER`, etc.
+* Xamarin
+    - This covers the existing Xamarin offerings, the new .NET 6-based iOS and
+      Android support is handled by the previous sections.
+    - The existing Xamarin Apple platforms don't have versioned preprocessor
+      symbols, but Android does: `__ANDROID_16__`
+    - We don't plan on adding `OR_GREATER` flavors like
+      `__ANDROID_16_OR_GREATER__`. Instead, we're only doing those for .NET 6.
```

Comment:
Make it clearer that there will be new symbols for Android/iOS:

```suggestion
      `__ANDROID_16_OR_GREATER__`. Instead, we're only adding the symbols described in the previous section for .NET 6.
```

This effectively means we'll be adding `ANDROID_21_OR_GREATER` since the TFM will be `net6.0-android21` right?

Reply:
Syntax-wise, yes. But I'm not sure where we landed for the TFM versions of Android. AFAIK we weren't using API version but product versions, but that's unrelated to this. The defines would match whatever the TFM version is.

Reply:
We did land on using the API versions, I'll need to update the design doc :)